### PR TITLE
extmarks: mark sanity/madness

### PR DESCRIPTION
**Pull request by bfredl**
This is an ongoing effort to make the extmark/bufhl implementation more sane via methodological madness.

Currently both bufhl and extmarks are used a nested structure of having a kbtree over lines only and an "inner" array/tree for columns. The point was to do both line number and column adjustments effective separately, but this is really sub-optimal. We can make the implementation both simpler and more effecient by storing both lines and columns in a single tree, the general idea is outlined e.g. in this atom blog post https://blog.atom.io/2015/06/16/optimizing-an-important-atom-primitive.html but there are some differences, like atom's approach "collapses" points at the same position to a set, while this implementation aims to keep only a "flat" structure apart from the single tree. 

- [x] implement new "marktree" derived from both kbtree and atom's marker-index.
- [x] use if for bufhl
- [x] use it for extmarks
- [ ] merge extmark, bufhl and virt text API into one coherent whole. 
- [ ] support intersecting ranges like atom's implementation (needed for multi-line "extranges", needs more work) 
- [ ] support byte based update event (WIP, intersects the new position adjustment code)
- [ ] add `nvim_buf_set_text` to preserve extmarks on modified lines.

### CODE REVIEW

**Comment by bfredl** [comment]
Extmarks and bufhl have now been merged to one single implementation internally. Next step would be to figure out an unified API for getting/setting extmarks with decorations.

File: src/nvim/marktree.c [hunk 1]
```diff
@@ -0,0 +1,1157 @@
+#include <assert.h>
```
Comment (justinmk) [inline src/nvim/marktree.c]: would be good to have a module description here similar to `multiqueue.c`, also mentioning common operations like `marktree_itr/del_get`.

Anyways, really cool stuff!

**Review by justinmk — APPROVED** [review]

**Comment by bfredl** [comment]
Multiline decorations are now working, sorta. https://github.com/bfredl/neovim/commit/5316c4e13820d34ced0d3c404d97e59048dabc03

Though it will break if too many ranges/marks overlap, we need a bit more logic for that.

File: src/nvim/mark_extended_defs.h [hunk 2]
```diff
@@ -10,6 +10,7 @@ typedef struct {
 } VirtTextChunk;
 
 typedef kvec_t(VirtTextChunk) VirtText;
+#define VIRTTEXT_EMPTY ((VirtText)KV_INITIAL_VALUE)
```
Comment (justinmk) [inline src/nvim/mark_extended_defs.h]: do we normally name this `VIRTTEXT_INIT` ? (cf. `STRING_INIT` et al.)

Comment (bfredl) [reply src/nvim/mark_extended_defs.h]: no, this is used only for _value_ locations where virt_text is optional, and not as _variable_ initialization before starting to use the array (use KV_INITIAL_VALUE for that). variable `_INIT` defines like `STRING_INIT` does not have a type specifier like this.

**Comment by bfredl** [comment]
I cleaned up most of the essential TODO items. Note there is still a lot of missing functionality, but I think we should merge the work so far as a first step quite soon, as it will fix crashes and correctness issues with the current extmark API. I will dog-feed it for a few days to get a bit more of coverage.

Note: tests for some of the new adjustment code is missing. I plan to do a systematic effort when byte-based buffer updates are implemented fully, because the correctness there is overlapping (i e if the byte-based update is correct and fully precise, then extmark adjustment will also be).

**Review by justinmk — APPROVED** [review]

**Comment by bfredl** [comment]
merged. Some stuff is still a bit rough/not fully reviewed, but things will change and improve in coming PRs.

**Pull request merged**